modules/logging/formattedencoder.go

```diff
@@ -0,0 +1,147 @@
+package logging
+
+import (
+	"bytes"
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/buger/jsonparser"
+	"github.com/caddyserver/caddy/v2"
+	"github.com/caddyserver/caddy/v2/caddyconfig/caddyfile"
+	"go.uber.org/zap/buffer"
+	"go.uber.org/zap/zapcore"
+)
+
+func init() {
+	caddy.RegisterModule(FormattedEncoder{})
+}
+
+const commonLogFormat = `{http.common_log}`
+
+// FormattedEncoder allows the user to provide custom template for log prints. The
+// encoder builds atop the json encoder, thus it follows its message structure. The placeholders
+// are namespaced by the name of the app logging the message.
+type FormattedEncoder struct {
+	zapcore.Encoder `json:"-"`
+	LogEncoderConfig
+	Template string `json:"template,omitempty"`
+}
+
+func (FormattedEncoder) CaddyModule() caddy.ModuleInfo {
+	return caddy.ModuleInfo{
+		ID: "caddy.logging.encoders.formatted",
+		New: func() caddy.Module {
+			return &FormattedEncoder{
+				Encoder: new(JSONEncoder),
+			}
+		},
+	}
+}
+
+// Provision sets up the encoder.
+func (se *FormattedEncoder) Provision(ctx caddy.Context) error {
+	if se.Template == "" {
+		return fmt.Errorf("missing template for formatted log encoder")
+	}
+	se.Encoder = zapcore.NewJSONEncoder(se.ZapcoreEncoderConfig())
+	return nil
+}
+
+// Clone wraps the underlying encoder's Clone. This is
+// necessary because we implement our own EncodeEntry,
+// and if we simply let the embedded encoder's Clone
+// be promoted, it would return a clone of that, and
+// we'd lose our FormattedEncoder's EncodeEntry.
+func (se FormattedEncoder) Clone() zapcore.Encoder {
+	return FormattedEncoder{
+		Encoder:  se.Encoder.Clone(),
+		Template: se.Template,
+	}
+}
+
+// EncodeEntry partially implements the zapcore.Encoder interface.
+func (se FormattedEncoder) EncodeEntry(ent zapcore.Entry, fields []zapcore.Field) (*buffer.Buffer, error) {
+	repl := caddy.NewReplacer()
+	buf, err := se.Encoder.EncodeEntry(ent, fields)
+	if err != nil {
+		return nil, err
+	}
+	appName := strings.SplitN(ent.LoggerName, ".", 2)[0]
+	// set the vals in the replacer
+	err = jsonparser.ObjectEach(buf.Bytes(), visitor(appName, repl))
+	buf.Reset() // the buffer is only used to collect placeholders' values anyway
+	buf.Free()
+	if err != nil {
+		return nil, err
+	}
+
+	out := repl.ReplaceKnown(se.Template, "")
+	// Unescape escaped quotes
+	buf.AppendString(strings.Replace(out, `\"`, `"`, -1))
```

Comment:
@mholt, is there anything besides this that needs to be unescaped?

Reply:
Probably 😅 

I think I have a better idea though, lemme see if I can explain it clearly in the review.